### PR TITLE
Feature file service inclusion

### PR DESCRIPTION
**Pull request by iAKM**
- File upload feature implemented
- Dev testing is done with the dev system
- Unit tests are failing at the moment. Need to change the existing unit tests
-  Execution strategy is written in readme file

Additional ENV variables to be added:
UPLOAD_TOPIC (the topic to send to once the upload is completed)

### CODE REVIEW

File: src/middleware/token-validation-middleware.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Validates the token and sends the user_id in the req.body
+ * the user id is available as `req.user_id`
+ * @param req - Initial request
+ * @param res  - Supposed response (to be filled by others)
+ * @param next - Next function
+ */
+export async function tokenValidator(req: Request, res: Response, next: NextFunction) {
+
+    const approvedRoles = ["tdei_admin", "poc", "flex_data_generator"]; // Change it based on service
```
Comment (susrisha) [inline src/middleware/token-validation-middleware.ts]: last one I think should be `pathways_data_generator`

File: src/model/gtfs-pathways-upload-meta.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    collected_by!:string;
+
+    @Prop()
+    @IsISO8601()
+    @IsNotEmpty()
+    collection_date!: Date;
+
+    @Prop()
+    @IsNotEmpty()
+    tdei_service_id!:string;
```
Comment (susrisha) [inline src/model/gtfs-pathways-upload-meta.ts]: change to `tdei_station_id`

File: src/service/event-bus-service.ts [hunk 3]
```diff
@@ -134,6 +137,37 @@ export class EventBusService implements IEventBusServiceInterface {
                 onError: this.processUploadError
             });
     }
+
+     /**
+     * Publishes the upload of a gtfs-flex file
+     */
+     public publishUpload(request:GtfsPathwaysUploadMeta, recordId:string,file_upload_path:string, userId:string, meta_file_path:string){
+        const messageContent =  QueueMessageContent.from({
+             stage:'flex-upload',
```
Comment (susrisha) [inline src/service/event-bus-service.ts]: stage name should be `pathways-upload`

File: src/service/event-bus-service.ts [hunk 4]
```diff
@@ -134,6 +137,37 @@ export class EventBusService implements IEventBusServiceInterface {
...
+                 'meta_file_path':meta_file_path
+             },
+             response:{
+                 success:true,
+                 message:'File uploaded for the organization: '+request.tdei_org_id+' with record id'+recordId
+             }
+         });
+         const message = QueueMessage.from(
+             {
+                 messageType:'gtfs-flex-upload',
```
Comment (susrisha) [inline src/service/event-bus-service.ts]: messageType shall be `pathways-upload`

File: src/service/storage-service.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    /**
+     * Upload the file to storage
+     * @param filePath File path of the file
+     * @param type mimetype of the file
+     * @param body Readable stream of the body
+     * @param containerName Name of the container. defaults to gtfs-flex
+     */
+    async uploadFile(filePath:string, type:string = 'application/zip' ,body:NodeJS.ReadableStream, containerName:string = 'gtfsflex'){
```
Comment (susrisha) [inline src/service/storage-service.ts]: container name is `gtfspathways`

File: test/gtfs-pathways-upload-meta-test.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import { validate } from "class-validator";
+import { GtfsPathwaysUploadMeta } from "../src/model/gtfs-pathways-upload-meta";
+import jwt, { Secret, JwtPayload } from 'jsonwebtoken';
+describe('GTFS Pathways Upload metadata test', ()=>{
+    describe('Unit tests', ()=>{
+
+        test('When supplied without collected by, meta validation should fail', async ()=>{
+            const meta = `{
+                "tdei_org_id":"5e339544-3b12-40a5-8acd-78c66d1fa981",
+                "tdei_service_id":"333",
```
Comment (susrisha) [inline test/gtfs-pathways-upload-meta-test.ts]: this is `tdei_station_id`

**Review by susrisha — CHANGES_REQUESTED** [review]
This is a bit mixed with flex and comments are written appropriately

File: src/middleware/token-validation-middleware.ts (hunk 1 above)
Comment (iAKM) [reply src/middleware/token-validation-middleware.ts]: fixed

File: src/model/gtfs-pathways-upload-meta.ts (hunk 2 above)
Comment (iAKM) [reply src/model/gtfs-pathways-upload-meta.ts]: fixed

File: README.md [hunk 7]
```diff
@@ -156,3 +156,141 @@ For running integration test, following env variables are required.
...
+### Add local server in pgadmin tool
+- go to http://localhost:5000 and add server with the following parameters
+
+- server name : Local
+- host: postgres
+- user: POSTGRES_USER in .env file
+- password: POSTGRES_PASSWROD in .env file
+
+### Import gtfs-flex database structure
+- In the sql query tool of the gtfs-flex database, execute the query available in `src/scripts/init.sql`
```
Comment (susrisha) [inline README.md]: gtfs-pathways database

File: src/controller/gtfs-pathways-controller.ts [hunk 8]
```diff
@@ -65,41 +99,56 @@ class GtfsPathwaysController implements IController {
...
+            const userId = request.body.user_id;
+            // Validate the meta data
+            const gtfsdto = GtfsPathwaysUploadMeta.from(meta);
+            const result = await validate(gtfsdto);
+            console.log('result', result);
+        
+            if(result.length != 0){
+                console.log('Metadata validation failed');
+                console.log(result);
+                // Need to send these as response
```
Comment (susrisha) [inline src/controller/gtfs-pathways-controller.ts]: Please check the other pull request for flex. The meta errors are sent back

File: test/common/tdei-object-faker.ts [hunk 9]
```diff
@@ -22,6 +23,20 @@ export class TdeiObjectFaker {
         } as PathwayVersions;
     }
 
+    static getGtfsPathwaysPayload(){
+        return {
+            polygon: this.getPolygon(),
+            tdei_org_id: 'tdei-org-id',
+            tdei_service_id:'tdei-service-id',
```
Comment (susrisha) [inline test/common/tdei-object-faker.ts]: tdei_station_id

**Review by susrisha — CHANGES_REQUESTED** [review]

File: src/service/event-bus-service.ts (hunk 3 above)
Comment (iAKM) [reply src/service/event-bus-service.ts]: fixed

File: src/service/event-bus-service.ts (hunk 4 above)
Comment (iAKM) [reply src/service/event-bus-service.ts]: fixed

File: src/service/storage-service.ts (hunk 5 above)
Comment (iAKM) [reply src/service/storage-service.ts]: fixed

File: README.md (hunk 7 above)
Comment (iAKM) [reply README.md]: fixed

File: src/controller/gtfs-pathways-controller.ts (hunk 8 above)
Comment (iAKM) [reply src/controller/gtfs-pathways-controller.ts]: fixed

File: test/common/tdei-object-faker.ts (hunk 9 above)
Comment (iAKM) [reply test/common/tdei-object-faker.ts]: fixed

**Review by susrisha — APPROVED** [review]
Seems correct. Accepting the PR

**Pull request merged**